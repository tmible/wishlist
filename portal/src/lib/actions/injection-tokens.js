/** @module Токены внедрения адаптеров действий */

/**
 * Токен внедрения сетевого сервиса
 * @constant {string}
 */
export const NetworkService = 'actions network service';
