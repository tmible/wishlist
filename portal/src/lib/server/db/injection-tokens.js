/** @module Токены внедрения адаптера БД */

/**
 * Токен внедрения БД
 * @constant {string}
 */
export const Database = 'wishlist database';

/**
 * SQL выражение для посчёта изменений
 * @constant {string}
 */
export const ChangesStatement = 'changes statement';
