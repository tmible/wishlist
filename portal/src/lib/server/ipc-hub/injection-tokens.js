/** @module Токены внедрения адаптера IPC хаба */

/**
 * Токен внедрения IPC хаба
 * @constant {string}
 */
export const IPCHub = 'IPC hub';
