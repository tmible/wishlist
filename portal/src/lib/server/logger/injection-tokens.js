/** @module Токены внедрения адаптера логирования */

/**
 * Токен внедрения логгера
 * @constant {string}
 */
export const Logger = 'logger';
