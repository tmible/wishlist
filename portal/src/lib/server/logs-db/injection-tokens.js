/** @module Токены внедрения адаптера БД с логами */

/**
 * Токен внедрения БД
 * @constant {string}
 */
export const Database = 'logs database';
