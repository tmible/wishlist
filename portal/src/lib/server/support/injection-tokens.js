/** @module Токены внедрения адаптеров поддержки */

/**
 * Токен внедрения кэша статусов доставки сообщений в поддержку
 * @constant {{}}
 */
export const MessageDeliveryStatus = {};
