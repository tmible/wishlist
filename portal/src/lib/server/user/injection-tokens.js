/** @module Токены внедрения потзователя */

/**
 * Токен внедрения серивиса работы с JWT
 * @constant {string}
 */
export const JWTService = 'user JWT service';
