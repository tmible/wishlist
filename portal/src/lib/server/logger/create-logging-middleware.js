import { inject } from '@tmible/wishlist-common/dependency-injector';
import {
  UNKNOWN_USER_UUID_COOKIE_NAME,
} from '$lib/constants/unknown-user-uuid-cookie-name.const.js';
import { Logger } from './injection-tokens.js';

/** @typedef {import('@sveltejs/kit').RequestEvent} RequestEvent */
/**
 * @template T
 * @typedef {import('@sveltejs/kit').MaybePromise<T>} MaybePromise
 * @see {@link https://svelte.dev/docs/kit/@sveltejs-kit#MaybePromise}
 */

/** @module Создание промежуточного обработчика, логирующего запросы и ответы на них */

/**
 * @typedef {Function} LoggingMiddleware
 * @param {RequestEvent} event Событие запроса
 * @param {() => MaybePromise<Response>} next Функция вызова следующих обработчиков
 * @returns {Promise<Response>} Ответ от следующих обработчиков
 * @async
 */

/**
 * Создание промежуточного обработчика
 * @function createLoggingMiddleware
 * @returns {LoggingMiddleware} Промежуточный обработчик
 */
export const createLoggingMiddleware = () => {
  const logger = inject(Logger);
  return async (event, next) => {
    const requestUuid = crypto.randomUUID();
    event.locals.requestUuid = requestUuid;

    logger.info(
      { requestUuid, unknownUserUuid: event.cookies.get(UNKNOWN_USER_UUID_COOKIE_NAME) ?? null },
      `request ${
        event.request.method
      } ${
        event.url.pathname
      }; cookie: ${
        event.cookies.getAll().map(({ name, value }) => `${name}=${value}`).join(';')
      }; body: ${
        event.request.body === null ? null : await event.request.clone().text()
      }`,
    );

    const response = await next(event);

    let responseBody = null;
    if (response.body !== null) {
      responseBody =
        response.body instanceof ReadableStream ?
          'event-stream' :
          (await response.clone().text());
    }

    logger.info(
      {
        requestUuid,
        unknownUserUuid: event.cookies.get(UNKNOWN_USER_UUID_COOKIE_NAME) ?? null,
        userid: event.locals.userid ?? null,
      },
      `response ${response.status}; body: ${responseBody}`,
    );

    return response;
  };
};
