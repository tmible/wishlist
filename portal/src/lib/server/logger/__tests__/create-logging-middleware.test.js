import { inject } from '@tmible/wishlist-common/dependency-injector';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { createLoggingMiddleware } from '../create-logging-middleware.js';
import { Logger } from '../injection-tokens.js';

vi.mock('@tmible/wishlist-common/dependency-injector');

describe('logger / create logging middleware', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should inject logger', () => {
    createLoggingMiddleware();
    expect(vi.mocked(inject)).toHaveBeenCalledWith(Logger);
  });

  it('should return middleware', () => {
    expect(createLoggingMiddleware()).toEqual(expect.any(Function));
  });

  describe('middleware', () => {
    const logger = { info: vi.fn() };
    const next = vi.fn();

    let middleware;
    let event;
    let response;

    beforeEach(() => {
      event = {
        locals: {},
        cookies: {
          get: vi.fn().mockReturnValue('unknown user UUID'),
          getAll: vi.fn(() => [{ name: 'name', value: 'value' }]),
        },
        request: {
          method: 'method',
          body: 'body',
          clone: vi.fn().mockReturnValue({ text: vi.fn().mockResolvedValue('body') }),
        },
        url: { pathname: 'pathname' },
      };
      response = {
        status: 'status',
        body: null,
      };
      vi.mocked(inject).mockReturnValueOnce(logger);
      vi.spyOn(crypto, 'randomUUID').mockReturnValueOnce('random UUID');
      next.mockReturnValueOnce(response);
      middleware = createLoggingMiddleware();
    });

    it('should assign request random UUID', async () => {
      await middleware(event, next);
      expect(event.locals.requestUuid).toBe('random UUID');
    });

    it('should log request', async () => {
      let info;
      logger.info.mockImplementationOnce((...args) => info = args);
      await middleware(event, next);
      expect(info).toMatchSnapshot();
    });

    it('should log request alternatively', async () => {
      event.cookies.get.mockReturnValueOnce(null);
      event.request.body = null;
      event.locals.userid = 'userid';
      let info;
      logger.info.mockImplementationOnce((...args) => info = args);
      await middleware(event, next);
      expect(info).toMatchSnapshot();
    });

    it('should call next', async () => {
      await middleware(event, next);
      expect(next).toHaveBeenCalledWith(event);
    });

    it('should log response', async () => {
      let info;
      logger.info.mockImplementationOnce(() => {}).mockImplementationOnce((...args) => info = args);
      await middleware(event, next);
      expect(info).toMatchSnapshot();
    });

    it('should log response alternatively', async () => {
      event.cookies.get.mockReturnValueOnce(null);
      event.request.body = null;
      event.locals.userid = 'userid';
      let info;
      logger.info.mockImplementationOnce(() => {}).mockImplementationOnce((...args) => info = args);
      await middleware(event, next);
      expect(info).toMatchSnapshot();
    });

    it('should return response', async () => {
      await expect(middleware(event, next)).resolves.toEqual(response);
    });
  });
});
