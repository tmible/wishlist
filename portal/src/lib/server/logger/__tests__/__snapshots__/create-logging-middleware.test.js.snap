// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger / create logging middleware > middleware > should log request 1`] = `
[
  {
    "requestUuid": "random UUID",
    "unknownUserUuid": "unknown user UUID",
  },
  "request method pathname; cookie: name=value; body: body",
]
`;

exports[`logger / create logging middleware > middleware > should log request alternatively 1`] = `
[
  {
    "requestUuid": "random UUID",
    "unknownUserUuid": null,
  },
  "request method pathname; cookie: name=value; body: null",
]
`;

exports[`logger / create logging middleware > middleware > should log response 1`] = `
[
  {
    "requestUuid": "random UUID",
    "unknownUserUuid": "unknown user UUID",
    "userid": null,
  },
  "response status; body: null",
]
`;

exports[`logger / create logging middleware > middleware > should log response alternatively 1`] = `
[
  {
    "requestUuid": "random UUID",
    "unknownUserUuid": "unknown user UUID",
    "userid": "userid",
  },
  "response status; body: null",
]
`;
